[ADD] hr-contract-cae: amendments

mclaeysb opened the pull request:
Adds 'contract amendments':

* Introduces two echelons of contract types: main contracts (CDI, CDD, ...) and amendments (Bonus, Salary Evolution, ...).
* A button is added to create an amendment of a contract. It's information, or that of it's latest amendment, is copied into a new amendment.
* Amendments of a main contract are indexed.
* Creation of amendments can be limited by type of amendment to a certain number
* Fields are added to contracts regarding duration and dates (signature, mailing, ...).
* Contract attachements can be uploaded

mclaeysb committed:
[ADD] hr-contract-cae WIP: add line of contract ammendments and extra date fields, and contract type echelon and max_usage

robinkeunen commented on hr_contract_cae/models/hr_contract.py:
```diff
@@ -1,11 +1,82 @@
+        default="main",
+    )
+    max_usage = fields.Integer(
+        string="Maximum Usage", default=1, required=True
```
I'd default to False, and make it not required. We don't have a limit on many contract types.

robinkeunen commented on hr_contract_cae/models/hr_contract.py:
```diff
@@ -1,11 +1,82 @@
+    latest_contract_id = fields.Many2one(
+        comodel_name="hr.contract",
+        string="Latest Contract",
+        readonly=True,
```
computed fields are necessarily readonly

robinkeunen commented on hr_contract_cae/models/hr_contract.py:
```diff
@@ -18,21 +89,108 @@ class Contract(models.Model):
         track_visibility="onchange",
         help="Employee's monthly gross wage.",
         compute="_compute_wage",
+        copy=False,
```
I don't think a copy with a (non stored) computed field makes sense.

robinkeunen commented on hr_contract_cae/models/hr_contract.py:
```diff
@@ -18,21 +89,108 @@ class Contract(models.Model):
+                        (contract.type_id.name or ""),
+                    ],
+                )
+            )
```
This is not human readable :-)

robinkeunen commented on hr_contract_cae/models/hr_contract.py:
```diff
@@ -18,21 +89,108 @@ class Contract(models.Model):
+        contract = latest_contract_id.copy(
+            {
+                "state": "draft",
+                "type_id": self.env["hr.contract.type"]
```
I'd leave it to False instead of embedding a sql request here.

mclaeysb replied:
That was not possible since the copy() wants to create the object, and the field is required...

robinkeunen commented on hr_contract_cae/models/hr_contract.py:
```diff
@@ -18,21 +89,108 @@ class Contract(models.Model):
+        }
+
+    @api.multi
+    @api.onchange("initial_contract_id", "type_id")
```
How can you have an `onchange` on a readonly field?

mclaeysb replied:
The readonly field could technically still be changed by a function. I didn't do this conciously though, I just looked at all the fields accessed in the code below and mentionned them here... :)

robinkeunen commented on hr_contract_cae/models/hr_contract.py:
```diff
@@ -1,11 +1,82 @@
+        copy=False,
+        compute="_compute_type_echelon",
+    )  # Todo: add translation "Echelone du Type de Contrat"
+    count_type = fields.Integer(
```
I'd use the more explicit `contract_type_count`.

mclaeysb replied:
Good point!

robinkeunen commented on hr_contract_cae/models/hr_contract.py:
```diff
@@ -18,21 +89,108 @@ class Contract(models.Model):
+            )
+            current_contract_id = current_contract_id.child_contract_id
+        self.latest_contract_id = current_contract_id
+        self.count_type = count_type
```
`count_type`: you're assigning a boolean to an integer and adding boolean.

mclaeysb replied:
Would this give errors or is it just non-zen-of-Python?
```python
a = 2
a += True
# a = 3
# seems to work...
```

robinkeunen replied:
it's not pythonesque according to me. When thinking about code, **arithmetic** operations on integer should give integer and operation on boolean should give booleans.

mclaeysb replied:
so you'd prefer to cast to boolean first?

robinkeunen replied:
if you're doing addition i'd rather cast to integers. Or more simply:

```python
while current_contract_id.child_contract_id:
    if current_contract_id.child_contract_id.type_id == self.type_id:
        count_type += 1
```

robinkeunen reviewed (commented):
Good start Manuel, I left a few comments hereunder. :-) On en parle lundi!

mclaeysb committed:
[REF] hr_contract: renaming, removing unlogical copy statements, echelon max_usage not required
renaming: count_type > contract_type_count, index_amendment > contract_amendment_index

mclaeysb committed:
[REF] respect 79 line length per OCA v12 guidelines

mclaeysb committed:
[ADD] hr_contract: attachements

mclaeysb committed:
[ADD] hr_contract_cae: check contract_type_count (WIP), no duration for amendments, rename contract_amendment_index

mclaeysb committed:
[ADD] hr_contract_cae: tree view of amendments

mclaeysb committed:
[ADD] hr_contract_cae: separate menu item for all, main and amendment contracts, and fix for all_contract_id error preventing contract creation

mclaeysb committed:
[FIX] hr_contract_cae: fix checking of max_usage

mclaeysb committed:
[FIX] hr_contract_cae: fix missing default type_echolon value
This was preventing a choice menu for type_id

mclaeysb committed:
[ADD] hr_contract_cae: restrict employee editing for amendments

mclaeysb committed:
[ADD] hr_contract_cae: add unidecode to requirements.txt

mclaeysb committed:
[ADD] hr_contract_cae: test of contract and amendment creation

mclaeysb committed:
[REF] hr_contract_cae: use contract_type_group

mclaeysb committed:
[REF] hr_contract_cae: reorder things

mclaeysb committed:
[ADD] hr_contract_cae: initial starting date

mclaeysb committed:
[REF] hr_contract_cae: move create_amendment to hr_contract

mclaeysb committed:
[IMP] hr_contract_cae: improve readme

mclaeysb committed:
[FIX] member_data_history: fix missing employee_id in test

mclaeysb commented:
@robinkeunen 
- I've added a commit hoping to the Travis CI by adding a line in the `member_data_history` module, but it might have been better to do this in that branch and/or merge both?
- There's one 'reordering' commit which might make reviewing a litter harder. It might make it easier for you to only visualise the diffs up until that commit.

mclaeysb commented:
(Did `git rebase` and `git push --force`, so the comments and commit's switched places a bit in this list)

robinkeunen commented on hr_contract_cae/models/hr_contract.py:
```diff
@@ -1,13 +1,109 @@
+        string="Contracts",
+    )
+
+    def get_initial(self):
+        return self.contract_ids.search(
+            [("contract_group_id", "=", self.id)], limit=1
+        )
```
I'd do a computed field and (maybe) store the result but a function is fine too.

robinkeunen commented on hr_contract_cae/models/hr_contract.py:
```diff
@@ -1,13 +1,109 @@
+                ("contract_group_id", "=", self.id),
+                ("amendment_index", "<", amendment_index),
+            ],
+            order="amendment_index desc",
```
[TIL](https://www.urbandictionary.com/define.php?term=TIL)

robinkeunen commented on hr_contract_cae/models/hr_contract.py:
```diff
@@ -1,13 +1,109 @@
+_logger = logging.getLogger(__name__)
+
+
+class ContractGroup(models.Model):
```
Much much cleaner/simpler, great refactoring 👍

robinkeunen commented on hr_contract_cae/models/hr_contract.py:
```diff
@@ -1,12 +1,137 @@
     _inherit = "hr.contract"
 
-    reason = fields.Char(string="Reason for recourse to contract")
+    name = fields.Char(compute="_compute_name")
```
You should store the name because
- it is very regularly accessed
  - besides avoiding recomputation, it will (i'm 75% sure) allow caching the name
- it will be useful to have the name in the database

robinkeunen commented on hr_contract_cae/models/hr_contract.py:
```diff
@@ -1,12 +1,137 @@
+        copy=False,
+        compute="_compute_type_echelon",
+        store=True,
+    )  # Todo: add translation "Echelone du Type de Contrat"
```
échelon

robinkeunen commented on hr_contract_cae/models/hr_contract.py:
```diff
@@ -1,12 +1,137 @@
+    parent_contract_id = fields.Many2one(
+        comodel_name="hr.contract",
+        string="Previous Contract",
+        ondelete="cascade",  # Deletion of the parent deletes the amendment
```
I'm not sure how that behaves on a computed field.

robinkeunen commented on hr_contract_cae/models/hr_contract.py:
```diff
@@ -1,12 +1,137 @@
+    contract_group_contract_ids = fields.One2many(
+        comodel_name="hr.contract", related="contract_group_id.contract_ids"
+    )
+    initial_contract_id = fields.Many2one(
+        comodel_name="hr.contract",
+        string="Initial Contract",
+        readonly=True,
+        compute="_compute_initial",
+    )
```
Option A: Are all these fields needed on the contract after your refactoring?
Option B: Shouldn't they rather be related fields on contract_group_contract_ids?

robinkeunen commented on hr_contract_cae/models/hr_contract.py:
```diff
@@ -1,12 +1,137 @@
+    child_contract_id = fields.Many2one(
+        comodel_name="hr.contract",
+        string="Next Contract",
+        readonly=True,
+        compute="_compute_child",
```
unstored computed field are already readonly

robinkeunen commented on hr_contract_cae/models/hr_contract.py:
```diff
@@ -1,12 +1,137 @@
+        compute="_compute_latest",
+    )
+
+    date_signature = fields.Date(
```
`signature_date` is better english.

robinkeunen commented on hr_contract_cae/models/hr_contract.py:
```diff
@@ -19,20 +144,162 @@ class Contract(models.Model):
+            contract_group = self.env["hr.contract.group"].create({})
+            res.contract_group_id = contract_group
+        if res.amendment_index == 0:
+            res.initial_contract_id = res.get_initial()
```
Would not bee needed if the `initial_contract_id` field was defined on the group. Plus, if `res.amendment_index == 0`, you already know that `res.initial_contract_id = self`.

robinkeunen commented on hr_contract_cae/models/hr_contract.py:
```diff
@@ -19,20 +144,162 @@ class Contract(models.Model):
+    @api.multi
+    @api.depends("employee_id", "initial_contract_id", "type_id")
+    def _compute_name(self):
+        for contract in self:
+            contract.name = "/".join(
+                filter(
+                    None,
+                    [
+                        (contract.employee_id.name or ""),
+                        (
+                            contract.initial_contract_id.type_id.name
+                            if contract.amendment_index > 0
+                            else ""
+                        ),
+                        (
+                            str(contract.amendment_index)
+                            if contract.amendment_index > 0
+                            else ""
+                        ),
+                        (contract.type_id.name or ""),
+                    ],
+                )
+            )
```
This is still very hard to read, I would write:

```python
for contract in self:
    employee_name = contract.employee_id.name or ""

    if contract.amendment_index > 0:
        contract_name = contract.initial_contract_id.type_id.name
        index = str(contract.amendment_index)
    else:
        contract_name = ""
        index = ""

    type_ = contract.type_id.name or ""

    contract.name = "/".join(
        filter(
            None,
            [
                employee_name,
                contract_name,
                index,
                type_
            ],
        )
    )
```

robinkeunen commented on hr_contract_cae/models/hr_contract.py:
```diff
@@ -19,20 +144,162 @@ class Contract(models.Model):
+            for c in contract.contract_group_contract_ids:
+                if c.type_id == contract.type_id:
+                    type_count += 1
+            contract.type_count = type_count
```
This feels like it should be on the group. Plus, now that you don't have to traverse the contract chain, you could do:

```python
contract.type_count = len(list(filter(
    λ c: c.type_id == contract.type_id,
    contract.contract_group_contract_ids)))
# or
contract.type_count = len([c for c in contract.contract_group_contract_ids
                           if c.type_id == contract.type_id])
```

robinkeunen commented on hr_contract_cae/models/hr_contract.py:
```diff
@@ -19,20 +144,162 @@ class Contract(models.Model):
+
+    def _compute_latest(self):
+        for contract in self:
+            contract.latest_contract_id = contract.get_latest()
```
I would put all of this in the group and use related fields here.

mclaeysb replied:
I don't think that would work because child and parent are function of the current contract, so can't be a property of the group.

robinkeunen commented on hr_contract_cae/models/hr_contract.py:
```diff
@@ -19,20 +144,162 @@ class Contract(models.Model):
+        return self.contract_group_id.get_child(self.amendment_index)
+
+    def get_latest(self):
+        return self.contract_group_id.get_latest()
```
There are too many layers here :-)

mclaeysb replied:
This layer could go indeed and be implemented in the `_compute` functions. These getters are of no use since the values are stored in the fields...

robinkeunen commented on hr_contract_cae/models/hr_contract.py:
```diff
@@ -1,8 +1,12 @@
 from odoo import _, api, fields, models
 from odoo.exceptions import ValidationError
 
+_logger = logging.getLogger(__name__)
```
👍

robinkeunen commented on hr_contract_cae/models/hr_contract.py:
```diff
@@ -86,6 +90,11 @@ class Contract(models.Model):
     reason = fields.Char(string="Reason for recourse to contract", copy=False)
 
     duration = fields.Integer(string="Duration", default=6)
+    date_initial_start = fields.Date(
```
`initial_start_date` is better english

mclaeysb replied:
This was for symmetry with `date_start` and `date_end` in `hr.contract`

robinkeunen commented on member_data_history/tests/test_value_log.py:
```diff
@@ -8,10 +8,17 @@
+    def setUp(self):
+
+        super(TestValueLog, self).setUp()
+
```
Too many newlines here :-)

robinkeunen reviewed (changes requested):
Nice work, I especially liked the refactoring you did on the groups. Smart. I think more things should be move to groups as a consequence. And groups will therefore deserve their own file.

I did not review xml and tests (but glad you wrote some)

Carry on, last mile!

robinkeunen commented on hr_contract_cae/__manifest__.py:
```diff
@@ -12,7 +12,12 @@
     "license": "AGPL-3",
     "external_dependencies": {"python": ["dateutil.relativedelta"]},
     "depends": ["hr", "hr_contract"],
```
You should add this module in the dependencies of `coopaname_custom`

mclaeysb committed:
[IMP] hr_cae_contract: required by coopaname_custome

mclaeysb committed:
[ADD] hr_cae_contract: add tags, refactor getters

mclaeysb commented:
Manipulations and feedback implemented. 
Let's rename `hr_contract_cae` > `hr_cae_contract` when we merge them all.
